### PR TITLE
[webnn] Add float16 tests for WebNN concat op

### PR DESCRIPTION
**Pull request by BruceDai**
I'm still working on implement scripts to generate test data & baseline data for float16 tests based on previous test data & baseline data of float32 precision. 
Here I firstly submit this PR to show the overview of testing float16 precision. 
These new added float16 tests in [webnn/resources/test_data/concat.json](https://github.com/web-platform-tests/wpt/compare/master...BruceDai:add_webnn_fp16_tests?expand=1#diff-5bb4136584e2de418ca029a0152669c15de000b5b4c57f7f2c008bf1aa7ae6c7) use the same input data of float64 precision as ones used for testing float32 precision before, and such expected baseline data of float16 precision. 

I'm struggled on readability like Solution-1 which is usingwith this PR and simple like Solution-2.

Solution 1 - positive: do not need modify more current test framework / negative: exist many duplicated test data code
```json
    {
      "name": "concat two float32 1D tensors of same shape along axis 0",
      "inputs": [
        {
          "name": "input1",
          "shape": [12],
          "data": [
            -0.39444134019222243,
            ... ...
            -0.6731740531810844
          ],
          "type": "float32"
        },
        {
          "name": "input2",
          "shape": [12],
          "data": [
            0.4918989118791477,
            ... ...
            0.1211843166661235
          ],
          "type": "float32"
        }
      ],
      "axis": 0,
      "expected": {
        "name": "output",
        "shape": [24],
        "data": [
          -0.3944413363933563,
          ... ...
          0.1211843192577362
        ],
        "type": "float32"
      }
    },
    {
      "name": "concat two float16 1D tensors of same shape along axis 0",
      "inputs": [
        {
          "name": "input1",
          "shape": [12],
          "data": [
            -0.39444134019222243,
           ... ...
            -0.6731740531810844
          ],
          "type": "float16"
        },
        {
          "name": "input2",
          "shape": [12],
          "data": [
            0.4918989118791477,
            ... ...
            0.1211843166661235
          ],
          "type": "float16"
        }
      ],
      "axis": 0,
      "expected": {
        "name": "output",
        "shape": [24],
        "data": [
          -0.39453125,
          ... ....
          0.12115478515625
        ],
        "type": "float16"
      }
    },
```

Solution 2 - positive: simple, no duplicated test data code  / negative: need modify more to support this updated struct, for examples,  need iterate precision keys list of "expected" dictionary to test float32 and float16 precisions, according to precision key to prepare relative TypedArray data for input/constant operands and output operand(s),  generate each test name with precision word for them to show clearly on UI, etc., 
```json
    {
      "name": "concat two 1D tensors of same shape along axis 0",
      "inputs": [
        {
          "name": "input1",
          "shape": [12],
          "data": [
            -0.39444134019222243,
            ... ...
            -0.6731740531810844
          ],
          "type": "float64"
        },
        {
          "name": "input2",
          "shape": [12],
          "data": [
            0.4918989118791477,
            ... ...
            0.1211843166661235
          ],
          "type": "float64"
        }
      ],
      "axis": 0,
      "expected": {
        "name": "output",
        "shape": [24],
        "data": {
           "float32": [ // baseline for test float32 precision
              -0.3944413363933563,
              ... ...
              0.1211843192577362
            ],
            "float16": [ // baseline for test float16 precision
              -0.39453125,
              ... ....
              0.12115478515625
            ]
         }
      }
    },
```

@fdwr PTAL, and any suggestions, thanks.

### CODE REVIEW

File: webnn/resources/utils.js [hunk 1]
```diff
@@ -662,4 +685,51 @@ const testWebNNOperation = (operationName, buildFunc) => {
...
+};
+
+// ref: http://stackoverflow.com/questions/32633585/how-do-you-convert-to-half-floats-in-javascript
+const toHalf = (value) => {
+  let floatView = new Float32Array(1);
+  let int32View = new Int32Array(floatView.buffer);
+
+  /* This method is faster than the OpenEXR implementation (very often
+   * used, eg. in Ogre), with the additional benefit of rounding, inspired
+   * by James Tursa?s half-precision code. */
```
Comment (fdwr) [inline webnn/resources/utils.js:697]: Tursa's?

File: webnn/resources/utils.js [hunk 2]
```diff
@@ -4,13 +4,30 @@ const ExecutionArray = ['sync', 'async'];
...
   int32: Int32Array,
   uint32: Uint32Array,
   int8: Int8Array,
   uint8: Uint8Array,
 };
 
+const getTypedArrayData = (type, data) => {
+  let outData;
+  if (type === 'float16') {
+    // workaround to convert Float16 to Unit16
```
Comment (fdwr) [inline webnn/resources/utils.js:20]: Uint16

File: webnn/resources/utils.js (hunk 1 above)
Comment (BruceDai) [reply webnn/resources/utils.js]: Thanks, fixed.

File: webnn/resources/utils.js (hunk 2 above)
Comment (BruceDai) [reply webnn/resources/utils.js]: Thanks, fixed.

File: webnn/resources/utils.js [hunk 3]
```diff
@@ -390,8 +407,14 @@ const assert_array_approx_equals_ulp = (actual, expected, nulp, dataType, descri
...
+      if (dataType === 'float32') {
+        actualBitwise = getBitwise(actual[i], dataType);
+        expectedBitwise = getBitwise(expected[i], dataType);
+      } else if (dataType === 'float16') {
+        actualBitwise = actual[i];
+        // convert expected data of Float16 to Uint16
+        expectedBitwise = toHalf(expected[i]);
+      }
       distance = actualBitwise - expectedBitwise;
       distance = distance >= 0 ? distance : -distance;
```
Comment (fdwr) [inline webnn/resources/utils.js:419]: Isn't this just `distance = Math.abs(distance)`?

**Review by fdwr — APPROVED** [review]
👍

**Review by Honry — APPROVED** [review]
LGTM, thanks!

**Pull request merged**